File: server/lib/services/analytics_service.dart
import '../models/transaction.dart';
import '../store/json_store.dart';

class Filters {
  final String? startDate;
  final String? endDate;
  final String? account;
  final String? category;

  Filters({this.startDate, this.endDate, this.account, this.category});
}

List<CleanTransaction> _applyFilters(
    List<CleanTransaction> transactions, Filters filters) {
  var filtered = transactions;

  if (filters.startDate != null) {
    filtered = filtered.where((tx) => tx.date.compareTo(filters.startDate!) >= 0).toList();
  }
  if (filters.endDate != null) {
    filtered = filtered.where((tx) => tx.date.compareTo(filters.endDate!) <= 0).toList();
  }
  if (filters.account != null) {
    final acct = filters.account!.toLowerCase();
    filtered = filtered
        .where((tx) => tx.accountLabel.toLowerCase().contains(acct))
        .toList();
  }
  if (filters.category != null) {
    final cat = filters.category!.toLowerCase();
    filtered = filtered
        .where((tx) => tx.category.toLowerCase() == cat)
        .toList();
  }

  return filtered;
}

List<Map<String, dynamic>> getCategoryBreakdown(Filters filters) {
  final transactions = _applyFilters(loadCleanTransactions(), filters);

  final categories = <String, Map<String, dynamic>>{};
  for (final tx in transactions) {
    final cat = tx.category;
    categories.putIfAbsent(cat, () => {'count': 0, 'total_cents': 0});
    categories[cat]!['count'] = (categories[cat]!['count'] as int) + 1;
    categories[cat]!['total_cents'] =
        (categories[cat]!['total_cents'] as int) + tx.amount;
  }

  final entries = categories.entries.toList()
    ..sort((a, b) =>
        (a.value['total_cents'] as int).compareTo(b.value['total_cents'] as int));

  return entries
      .map((e) => {
            'category': e.key,
            'count': e.value['count'],
            'total_cents': e.value['total_cents'],
            'total':
                '\$${((e.value['total_cents'] as int).abs() / 100).toStringAsFixed(2)}',
          })
      .toList();
}

List<Map<String, dynamic>> getMonthlyBreakdown(Filters filters) {
  final transactions = _applyFilters(loadCleanTransactions(), filters);

  final months = <String, Map<String, dynamic>>{};
  for (final tx in transactions) {
    final key = '${tx.year}-${tx.month.toString().padLeft(2, '0')}';
    months.putIfAbsent(key, () => {'count': 0, 'total_cents': 0});
    months[key]!['count'] = (months[key]!['count'] as int) + 1;
    months[key]!['total_cents'] =
        (months[key]!['total_cents'] as int) + tx.amount;
  }

  final entries = months.entries.toList()
    ..sort((a, b) => a.key.compareTo(b.key));

  return entries
      .map((e) => {
            'month': e.key,
            'count': e.value['count'],
            'total_cents': e.value['total_cents'],
            'total':
                '\$${((e.value['total_cents'] as int).abs() / 100).toStringAsFixed(2)}',
          })
      .toList();
}

String _getWeekStart(String dateStr) {
  final d = DateTime.parse('${dateStr}T00:00:00Z');
  final day = d.weekday % 7; // Sunday = 0
  final weekStart = d.subtract(Duration(days: day));
  return weekStart.toIso8601String().substring(0, 10);
}

List<Map<String, dynamic>> getWeeklyBreakdown(Filters filters) {
  final transactions = _applyFilters(loadCleanTransactions(), filters);

  final weeks = <String, Map<String, dynamic>>{};
  for (final tx in transactions) {
    if (tx.date.isEmpty) continue;
    final week = _getWeekStart(tx.date);
    weeks.putIfAbsent(week, () => {'count': 0, 'total_cents': 0});
    weeks[week]!['count'] = (weeks[week]!['count'] as int) + 1;
    weeks[week]!['total_cents'] =
        (weeks[week]!['total_cents'] as int) + tx.amount;
  }

  final entries = weeks.entries.toList()
    ..sort((a, b) => a.key.compareTo(b.key));

  return entries
      .map((e) => {
            'week_start': e.key,
            'count': e.value['count'],
            'total_cents': e.value['total_cents'],
            'total':
                '\$${((e.value['total_cents'] as int).abs() / 100).toStringAsFixed(2)}',
          })
      .toList();
}

File: server/src/services/analyticsService.ts
import { loadCleanTransactions } from '../store/jsonStore.js';
import type { AnalyticsFilters, CleanTransaction } from '../types.js';

function applyFilters(transactions: CleanTransaction[], filters: AnalyticsFilters): CleanTransaction[] {
  // Only expenses (negative amounts), exclude Transfer category
  let filtered = transactions.filter((tx) => tx.amount < 0 && tx.category !== 'Transfer');

  if (filters.startDate) {
    filtered = filtered.filter((tx) => tx.date >= filters.startDate!);
  }
  if (filters.endDate) {
    filtered = filtered.filter((tx) => tx.date <= filters.endDate!);
  }
  if (filters.account) {
    const acct = filters.account.toLowerCase();
    filtered = filtered.filter((tx) => (tx.account_label ?? '').toLowerCase().includes(acct));
  }
  if (filters.category) {
    const cat = filters.category.toLowerCase();
    filtered = filtered.filter((tx) => tx.category.toLowerCase() === cat);
  }

  return filtered;
}

export function getCategoryBreakdown(filters: AnalyticsFilters): object[] {
  const transactions = applyFilters(loadCleanTransactions(), filters);

  const categories = new Map<string, { count: number; total_cents: number }>();
  for (const tx of transactions) {
    const cat = tx.category;
    if (!categories.has(cat)) categories.set(cat, { count: 0, total_cents: 0 });
    const entry = categories.get(cat)!;
    entry.count += 1;
    entry.total_cents += tx.amount;
  }

  const entries = [...categories.entries()].sort(
    (a, b) => a[1].total_cents - b[1].total_cents,
  );

  return entries.map(([category, { count, total_cents }]) => ({
    category,
    count,
    total_cents,
    total: `$${(Math.abs(total_cents) / 100).toFixed(2)}`,
  }));
}

export function getMonthlyBreakdown(filters: AnalyticsFilters): object[] {
  const transactions = applyFilters(loadCleanTransactions(), filters);

  const months = new Map<string, { count: number; total_cents: number }>();
  for (const tx of transactions) {
    const key = `${tx.year}-${String(tx.month).padStart(2, '0')}`;
    if (!months.has(key)) months.set(key, { count: 0, total_cents: 0 });
    const entry = months.get(key)!;
    entry.count += 1;
    entry.total_cents += tx.amount;
  }

  const entries = [...months.entries()].sort((a, b) => a[0].localeCompare(b[0]));

  return entries.map(([month, { count, total_cents }]) => ({
    month,
    count,
    total_cents,
    total: `$${(Math.abs(total_cents) / 100).toFixed(2)}`,
  }));
}

function getWeekStart(dateStr: string): string {
  const d = new Date(`${dateStr}T00:00:00Z`);
  // JS: Sunday=0, Monday=1, ... Saturday=6
  const day = d.getDay(); // Use getDay() directly (not % 7)
  const weekStart = new Date(d);
  weekStart.setUTCDate(d.getUTCDate() - day);
  return weekStart.toISOString().slice(0, 10);
}

export function getWeeklyBreakdown(filters: AnalyticsFilters): object[] {
  const transactions = applyFilters(loadCleanTransactions(), filters);

  const weeks = new Map<string, { count: number; total_cents: number }>();
  for (const tx of transactions) {
    if (!tx.date) continue;
    const week = getWeekStart(tx.date);
    if (!weeks.has(week)) weeks.set(week, { count: 0, total_cents: 0 });
    const entry = weeks.get(week)!;
    entry.count += 1;
    entry.total_cents += tx.amount;
  }

  const entries = [...weeks.entries()].sort((a, b) => a[0].localeCompare(b[0]));

  return entries.map(([week_start, { count, total_cents }]) => ({
    week_start,
    count,
    total_cents,
    total: `$${(Math.abs(total_cents) / 100).toFixed(2)}`,
  }));
}
